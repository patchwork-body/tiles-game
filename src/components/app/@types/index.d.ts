type AppComponent = (props: React.ComponentProps) => ReactElement;
