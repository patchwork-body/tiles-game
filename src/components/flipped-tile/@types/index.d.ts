type FlippedTileProps = {
  hidden: boolean;
  color: string;
  index: number;
  clickHandler(index: number): void;
};
type FlippedTileComponent = (props: FlippedTileProps) => ReactElement;
