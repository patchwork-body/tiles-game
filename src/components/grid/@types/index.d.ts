import { List, Map } from 'immutable';

type GridState = Map<{
  prevRevealedTiles: Map<number, GridTile>;
  tiles: List<List<GridTile>>;
}>;

type GridTile = { removed: boolean } & TileProps;

type GridAction = {
  type: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  payload: any;
};

type GridReducer = (state: GridState, action: GridAction) => GridState;

type GridProps = {
  rows: number;
  columns: number;
};

type GridComponent = (props: GridProps) => ReactElement;
