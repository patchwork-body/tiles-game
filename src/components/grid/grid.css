.grid {
  margin: 5rem auto;
  display: grid;
  gap: 2rem 1.5rem;
  width: fit-content;
}

.grid__tile {
  background-color: #c7c7c7;
  border-radius: 0.5rem;
  width: 6rem;
  min-height: 6rem;
}
