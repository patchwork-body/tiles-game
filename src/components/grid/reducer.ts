import { List, Map } from 'immutable';
import { GridReducer, GridState, GridTileData } from './@types';

enum GRID_ACTIONS {
  REVEAL_TILE = '@REVEAL_TILE',
  REMOVE_IDENTITY = '@REMOVE_IDENTITY',
  HIDE_ALL = '@HIDE_ALL',
}

const gridReducer: GridReducer = (state, action) => {
  if (action.type === GRID_ACTIONS.REVEAL_TILE) {
    const { row, column } = action.payload;

    const targetTile = {
      ...state.getIn(['tiles', row, column]),
      hidden: false,
    };

    return state
      .updateIn(['tiles', row, column], () => targetTile)
      .update('prevRevealedTiles', (prev: Map<number, GridTileData>) =>
        prev.set(targetTile.index, targetTile),
      );
  } else if (action.type === GRID_ACTIONS.REMOVE_IDENTITY) {
    const { result: hasIdentity } = state.get('prevRevealedTiles').reduce(
      (prev: { result: boolean; value: string }, tile: GridTileData) => ({
        result: prev.value === tile.color,
        value: tile.color,
      }),
      { result: false, value: '' },
    );

    if (hasIdentity) {
      return state
        .get('prevRevealedTiles')
        .reduce((state: GridState, revealedTile: GridTileData) => {
          return state.updateIn(
            ['tiles', revealedTile.row, revealedTile.column],
            (tile: GridTileData) => ({
              ...tile,
              removed: true,
            }),
          );
        }, state);
    } else {
      return state;
    }
  } else if (action.type === GRID_ACTIONS.HIDE_ALL) {
    return state
      .update('tiles', (tiles: List<List<GridTileData>>) =>
        tiles.map((row: List<GridTileData>) =>
          row.map(tile => ({ ...tile, hidden: true })),
        ),
      )
      .update('prevRevealedTiles', () => Map());
  } else {
    return state;
  }
};

export { GRID_ACTIONS, gridReducer };
