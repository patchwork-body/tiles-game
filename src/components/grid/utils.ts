import { List } from 'immutable';
import { GridTileData } from './@types';

const genRandomInt = (to = 256): number => Math.floor(Math.random() * to);

const genColorList = (size: number): List<string> => {
  const colors = new Array(Math.ceil(size / 2)).fill(null).map(() => {
    return `rgb(${genRandomInt()}, ${genRandomInt()}, ${genRandomInt()})`;
  });
  return List(colors.concat(colors)).sortBy(Math.random);
};

const genRandomTiles = ({
  rows,
  columns,
  colors,
}: {
  rows: number;
  columns: number;
  colors: List<string>;
}): List<GridTileData> => {
  let tiles = List();

  for (let rowIndex = 0; rowIndex < rows; rowIndex++) {
    let row = List();

    for (let columnIndex = 0; columnIndex < columns; columnIndex++) {
      const index = columns * rowIndex + columnIndex;

      row = row.push({
        row: rowIndex,
        column: columnIndex,
        removed: false,
        index,
        hidden: true,
        color: colors.get(index),
      });
    }

    tiles = tiles.push(row);
  }
  return tiles;
};

export { genRandomInt, genColorList, genRandomTiles };
