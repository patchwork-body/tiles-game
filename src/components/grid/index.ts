import './grid.css';

export * from './grid';
