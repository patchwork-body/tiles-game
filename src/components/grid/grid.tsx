import React, { useReducer } from 'react';
import { List, Map } from 'immutable';

import { Tile } from '../tile';
import { gridReducer } from './reducer';
import { GridComponent, GridTile } from './@types';

const Grid: GridComponent = ({ rows, columns }) => {
  const tiles = [];
  const colors = ['red', 'green', 'blue', 'pink'];

  for (let rowIndex = 1; rowIndex <= rows; rowIndex++) {
    const row = [];
    for (let columnIndex = 1; columnIndex <= columns; columnIndex++) {
      row.push({ hidden: true, color: colors[columnIndex - 1] });
    }
    tiles.push(List.of(...row));
  }

  const [state, dispatch] = useReducer(
    gridReducer,
    Map({ tiles: List.of(...tiles) }),
  );
  console.log(dispatch);

  return (
    <div className="grid">
      {state.get('tiles').map((row: List<GridTile>, rowIndex: number) => {
        return row
          .sortBy(Math.random)
          .map((tile: GridTile, columnIndex: number) => {
            return (
              <div
                className="grid__tile"
                key={rowIndex + columnIndex}
                style={{ gridRow: rowIndex + 1, gridColumn: columnIndex + 1 }}
              >
                <Tile {...tile} />
              </div>
            );
          });
      })}
    </div>
  );
};

export { Grid };
