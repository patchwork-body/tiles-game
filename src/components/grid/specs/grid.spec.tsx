import React from 'react';
import { render } from '@testing-library/react';

import { Grid } from '../grid';
import { GridProps } from '../@types';

describe('GridTile', () => {
  const rows = 4;
  const columns = 4;

  const props: GridProps = {
    rows,
    columns,
  };

  test('should match snapshot', () => {
    const { container } = render(<Grid {...props} />);
    expect(container.firstElementChild).toMatchSnapshot();
  });
});
