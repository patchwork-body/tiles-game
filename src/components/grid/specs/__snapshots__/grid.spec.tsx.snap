// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GridTile should match snapshot 1`] = `
<div
  class="grid"
>
  <div
    class="grid__tile"
    style="grid-row: 1; grid-column: 1; opacity: 1;"
  >
    <div
      class="flipped-tile flipped-tile_hidden"
      data-index="0"
      style="background: rgb(103, 216, 123);"
    />
  </div>
  <div
    class="grid__tile"
    style="grid-row: 1; grid-column: 2; opacity: 1;"
  >
    <div
      class="flipped-tile flipped-tile_hidden"
      data-index="1"
      style="background: rgb(189, 8, 26);"
    />
  </div>
  <div
    class="grid__tile"
    style="grid-row: 1; grid-column: 3; opacity: 1;"
  >
    <div
      class="flipped-tile flipped-tile_hidden"
      data-index="2"
      style="background: rgb(150, 69, 252);"
    />
  </div>
  <div
    class="grid__tile"
    style="grid-row: 1; grid-column: 4; opacity: 1;"
  >
    <div
      class="flipped-tile flipped-tile_hidden"
      data-index="3"
      style="background: rgb(68, 21, 24);"
    />
  </div>
  <div
    class="grid__tile"
    style="grid-row: 2; grid-column: 1; opacity: 1;"
  >
    <div
      class="flipped-tile flipped-tile_hidden"
      data-index="4"
      style="background: rgb(68, 21, 24);"
    />
  </div>
  <div
    class="grid__tile"
    style="grid-row: 2; grid-column: 2; opacity: 1;"
  >
    <div
      class="flipped-tile flipped-tile_hidden"
      data-index="5"
      style="background: rgb(177, 223, 153);"
    />
  </div>
  <div
    class="grid__tile"
    style="grid-row: 2; grid-column: 3; opacity: 1;"
  >
    <div
      class="flipped-tile flipped-tile_hidden"
      data-index="6"
      style="background: rgb(177, 223, 153);"
    />
  </div>
  <div
    class="grid__tile"
    style="grid-row: 2; grid-column: 4; opacity: 1;"
  >
    <div
      class="flipped-tile flipped-tile_hidden"
      data-index="7"
      style="background: rgb(163, 197, 37);"
    />
  </div>
  <div
    class="grid__tile"
    style="grid-row: 3; grid-column: 1; opacity: 1;"
  >
    <div
      class="flipped-tile flipped-tile_hidden"
      data-index="8"
      style="background: rgb(163, 197, 37);"
    />
  </div>
  <div
    class="grid__tile"
    style="grid-row: 3; grid-column: 2; opacity: 1;"
  >
    <div
      class="flipped-tile flipped-tile_hidden"
      data-index="9"
      style="background: rgb(119, 237, 222);"
    />
  </div>
  <div
    class="grid__tile"
    style="grid-row: 3; grid-column: 3; opacity: 1;"
  >
    <div
      class="flipped-tile flipped-tile_hidden"
      data-index="10"
      style="background: rgb(150, 69, 252);"
    />
  </div>
  <div
    class="grid__tile"
    style="grid-row: 3; grid-column: 4; opacity: 1;"
  >
    <div
      class="flipped-tile flipped-tile_hidden"
      data-index="11"
      style="background: rgb(189, 8, 26);"
    />
  </div>
  <div
    class="grid__tile"
    style="grid-row: 4; grid-column: 1; opacity: 1;"
  >
    <div
      class="flipped-tile flipped-tile_hidden"
      data-index="12"
      style="background: rgb(103, 216, 123);"
    />
  </div>
  <div
    class="grid__tile"
    style="grid-row: 4; grid-column: 2; opacity: 1;"
  >
    <div
      class="flipped-tile flipped-tile_hidden"
      data-index="13"
      style="background: rgb(119, 237, 222);"
    />
  </div>
  <div
    class="grid__tile"
    style="grid-row: 4; grid-column: 3; opacity: 1;"
  >
    <div
      class="flipped-tile flipped-tile_hidden"
      data-index="14"
      style="background: rgb(217, 160, 42);"
    />
  </div>
  <div
    class="grid__tile"
    style="grid-row: 4; grid-column: 4; opacity: 1;"
  >
    <div
      class="flipped-tile flipped-tile_hidden"
      data-index="15"
      style="background: rgb(217, 160, 42);"
    />
  </div>
</div>
`;
